src/frontend/src/handler/create_function.rs

```diff
@@ -172,3 +231,78 @@ pub async fn handle_create_function(
 
     Ok(PgResponse::empty_result(StatementType::CREATE_FUNCTION))
 }
+
+/// Upload wasm binary to object store.
+async fn upload_wasm_binary(
+    wasm_storage_url: &str,
+    object_name: &str,
+    wasm_binary: Bytes,
+) -> Result<()> {
+    // Note: it will panic if the url is invalid. We did a validation on meta startup.
+    let object_store = build_remote_object_store(
+        wasm_storage_url,
+        Arc::new(ObjectStoreMetrics::unused()),
+        "Wasm Engine",
+        ObjectStoreConfig::default(),
+    )
+    .await;
+    object_store
+        .upload(object_name, wasm_binary)
+        .await
+        .context("failed to upload wasm binary to object store")?;
+    Ok(())
+}
+
+/// Check if the function exists in the wasm binary.
+fn check_wasm_function(runtime: &arrow_udf_wasm::Runtime, identifier: &str) -> Result<()> {
+    if !runtime.functions().contains(&identifier) {
+        return Err(ErrorCode::InvalidParameterValue(format!(
+            "function not found in wasm binary: \"{}\"\nHINT: available functions:\n  {}",
+            identifier,
+            runtime.functions().join("\n  ")
+        ))
+        .into());
+    }
+    Ok(())
+}
+
+/// Generate the function identifier in wasm binary.
+fn wasm_identifier(name: &str, args: &[DataType], ret: &DataType, table_function: bool) -> String {
+    format!(
+        "{}({}){}{}",
+        name,
+        args.iter().map(datatype_name).join(","),
+        if table_function { "->>" } else { "->" },
+        datatype_name(ret)
+    )
+}
+
+/// Convert a data type to string used in identifier.
+fn datatype_name(ty: &DataType) -> String {
```

Comment:
Why didn't we use `DataType`'s `Display`? 🤪

Reply:
The type names defined in WASM module are slightly different than RW's type names, as the former follows Arrow specification while the latter follows postgres'.

Reply:
But I think arrow doesn't specified things like `T[]`, and thus the "string representation of a datatype" is still kind of arbitrary IMO.

So does the signature matter in the arrow-udf-wasm crate, or it can be decided by the caller?

Reply:
> But I think arrow doesn't specified things like `T[]`, and thus the "string representation of a datatype" is still kind of arbitrary IMO.

Sure. `T[]` is still pg style and should be changed to `list<T>` in my opinion.

> So does the signature matter in the arrow-udf-wasm crate, or it can be decided by the caller?

The signature format is a part of `arrow-udf` API and may vary between versions. (e.g. from API version 2 to 3, the type name of `bigint` was changed from `int8` to `int64`, in order to be consistent with Arrow and avoid ambiguity) The caller needs to use the correct name according to [the API version](https://docs.rs/arrow-udf/latest/arrow_udf/ffi/static.ARROWUDF_VERSION_3_0.html) declared by the WASM module.